### PR TITLE
VNSWRR limit the number of virtual peers when init

### PR DESCRIPTION
**Pull request by jizhuozhi**
#1716

### CODE REVIEW

**Comment by jizhuozhi** [comment]
hello, @chobits and @wangfakang , PLAT, thanks

File: modules/ngx_http_upstream_vnswrr_module/ngx_http_upstream_vnswrr_module.c [hunk 1]
```diff
@@ -159,7 +164,39 @@ ngx_http_upstream_vnswrr(ngx_conf_t *cf, ngx_command_t *cmd, void *conf)
...
                   |NGX_HTTP_UPSTREAM_DOWN;
 
+    uvnscf = ngx_http_conf_upstream_srv_conf(uscf,
+                                ngx_http_upstream_vnswrr_module);
+
+    value = cf->args->elts;
+
+    max_init = 0;
+
+    for (i = 2; i < cf->args->nelts; i++) {
```
[the remark this answers is not in the document]
Comment (jizhuozhi) [reply modules/ngx_http_upstream_vnswrr_module/ngx_http_upstream_vnswrr_module.c:174]: My mistake, it's copied from other command but some argument is skipped. Fixed.

File: modules/ngx_http_upstream_vnswrr_module/ngx_http_upstream_vnswrr_module.c [hunk 2]
```diff
@@ -159,7 +164,39 @@ ngx_http_upstream_vnswrr(ngx_conf_t *cf, ngx_command_t *cmd, void *conf)
...
+            max_init = ngx_atoi(&value[i].data[9], value[i].len - 9);
+
+            if (max_init == NGX_ERROR) {
+                goto invalid;
+            }
+
+            continue;
+        }
+
+        goto invalid;
```
[the remark this answers is not in the document]
Comment (jizhuozhi) [reply modules/ngx_http_upstream_vnswrr_module/ngx_http_upstream_vnswrr_module.c:187]: Keep a unified style with other nginx command parsing.

https://github.com/nginx/nginx/blob/master/src/http/ngx_http_upstream.c#L6153-L6168

File: modules/ngx_http_upstream_vnswrr_module/ngx_http_upstream_vnswrr_module.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (jizhuozhi) [reply modules/ngx_http_upstream_vnswrr_module/ngx_http_upstream_vnswrr_module.c]: Keep a unified style with other nginx command parsing. We can avoid dealing with optional parameters or a lot of modifications when adding new parameters

File: modules/ngx_http_upstream_vnswrr_module/ngx_http_upstream_vnswrr_module.c [hunk 3]
```diff
@@ -204,15 +233,24 @@ ngx_http_upstream_init_vnswrr(ngx_conf_t *cf,
...
+    max_init = uvnscf->max_init;
+
     uvnscf->init_number = NGX_CONF_UNSET_UINT;
     uvnscf->last_number = NGX_CONF_UNSET_UINT;
     uvnscf->last_peer = NULL;
     uvnscf->next = NULL;
     uvnscf->gcd = g;
 
+    if (!max_init) {
+        uvnscf->max_init = peers->number;
```
Comment (wangfakang) [inline modules/ngx_http_upstream_vnswrr_module/ngx_http_upstream_vnswrr_module.c:247]: style: need a blank before `else if {`. Same elsewhere.

File: modules/ngx_http_upstream_vnswrr_module/ngx_http_upstream_vnswrr_module.c [hunk 4]
```diff
@@ -238,6 +276,14 @@ ngx_http_upstream_init_vnswrr(ngx_conf_t *cf,
         ubvnscf->last_number = NGX_CONF_UNSET_UINT;
         ubvnscf->last_peer = NULL;
         ubvnscf->gcd = bg;
+        
+        if (!max_init) {
+            ubvnscf->max_init = backup->number;
+        } else if (max_init > backup->total_weight) {
```
Comment (wangfakang) [inline modules/ngx_http_upstream_vnswrr_module/ngx_http_upstream_vnswrr_module.c:282]: style.
https://tengine.taobao.org/book/appendix_a.html

File: modules/ngx_http_upstream_vnswrr_module/ngx_http_upstream_vnswrr_module.c [hunk 5]
```diff
@@ -238,6 +276,14 @@ ngx_http_upstream_init_vnswrr(ngx_conf_t *cf,
...
+            ubvnscf->max_init = backup->total_weight;
+        } else {
+            ubvnscf->max_init = max_init;
```
Comment (wangfakang) [inline modules/ngx_http_upstream_vnswrr_module/ngx_http_upstream_vnswrr_module.c:285]: use `ubvnscf->max_init = max_init;` for init and delete the `else`.

**Review by wangfakang — APPROVED** [review]

**Pull request merged**